backend/internal/user/service/userservice.go

```diff
@@ -112,118 +123,217 @@ func extractCredentials(user *model.User) (*model.Credentials, error) {
 	return &model.Credentials{}, nil
 }
 
-// GetUserList list the users.
-func (as *UserService) GetUserList() ([]model.User, error) {
-	users, err := store.GetUserList()
-	if err != nil {
+// GetUserList lists the users.
+func (as *UserService) GetUserList(limit, offset int) (*model.UserListResponse, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, loggerComponentName))
+
+	if err := validatePaginationParams(limit, offset); err != nil {
 		return nil, err
 	}
 
-	return users, nil
+	totalCount, err := store.GetUserListCount()
+	if err != nil {
+		logger.Error("Failed to get user list count", log.Error(err))
+		return nil, &constants.ErrorInternalServerError
+	}
+
+	users, err := store.GetUserList(limit, offset)
+	if err != nil {
+		logger.Error("Failed to get user list", log.Error(err))
+		return nil, &constants.ErrorInternalServerError
+	}
+
+	var links []model.Link
+	if offset+limit < totalCount {
+		links = append(links, model.Link{
+			Href: fmt.Sprintf("users?offset=%d&limit=%d", offset+limit, limit),
+			Rel:  "next",
+		})
+	}
+
+	response := &model.UserListResponse{
+		TotalResults: totalCount,
+		StartIndex:   offset + 1,
+		Count:        len(users),
+		Users:        users,
+		Links:        links,
+	}
+
+	return response, nil
 }
 
 // GetUser get the user for given user id.
-func (as *UserService) GetUser(userID string) (*model.User, error) {
+func (as *UserService) GetUser(userID string) (*model.User, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, loggerComponentName))
+	logger.Debug("Retrieving user", log.String("id", userID))
+
 	if userID == "" {
-		return nil, errors.New("user ID is empty")
+		return nil, &constants.ErrorMissingUserID
 	}
 
 	user, err := store.GetUser(userID)
 	if err != nil {
-		return nil, err
+		if errors.Is(err, constants.ErrUserNotFound) {
+			logger.Debug("User not found", log.String("id", userID))
+			return nil, &constants.ErrorUserNotFound
+		}
+		logger.Error("Failed to retrieve user", log.String("id", userID), log.Error(err))
+		return nil, &constants.ErrorInternalServerError
```

Comment:
Suggestion: We can extract this into a helper method as the code getting duplicated in multiple places